### ISSUE TITLE
Upgrading Debian Package Loses All Plugins

### ISSUE THREAD

**Issue by tebriel** [elastic/logstash#3540]
Upgrading by debian package loses all installed plugins.

## Steps to Reproduce ##

1. Install logstash via `dpkg -i logstash-1.5.1.deb`
1. Install a plugin `/opt/logstash/bin/plugin install my-awesome-filter.gem`
1. Upgrade logstash via `dpkg -i logstash-1.5.2.deb`
1. Try to start logstash with a configuration that requires 'my-awesome-filter', logstash will fail to start because it cannot find the filter.

Maybe this could be avoided by installing plugins to a separate directory? How does Elasticsearch handle this? Elasticsearch always preserves plugins on upgrade.

Thank you!!

**Comment by jordansissel** [comment]
I think this has to do with the deb package overwriting the Gemfile and Gemfile.lock during an upgrade. We'll want a way to help users upgrade without losing a record of all separately-installed plugins

**Comment by tebriel** [comment]
@jordansissel So fast! Thank you!

**Comment by jordansissel** [comment]
I think this also affects users of rpm and tar and zip artifacts in that the same thing will happen - the record of what plugins are installed can be overwritten and it'd be nice if there were a way to preserve that information.

**Comment by driegel** [comment]
@jordansissel YES, it affects the RPM and I believe that this is also the reason for issue #3201.
Please see my comment there.

**Comment by ThorbenJ** [comment]
I just upgraded LogStash from 5.3.0 to LogStash 5.3.2 and I've lost all (non-standard) plugins including x-pack!!

**Comment by jsmucr** [comment]
This is STILL an issue in 6.3.0.
